File: AsyncClientServerSSL/D2007/ClientSSL/Unit2.pas
unit Unit2;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, StdCtrls, clTcpClient, clSspiTls, clAsyncClient, clSocket, clWUtils;

type
  TForm2 = class(TForm)
    Label1: TLabel;
    edtServer: TEdit;
    Label2: TLabel;
    edtPort: TEdit;
    btnConnect: TButton;
    btnDisconnect: TButton;
    edtData: TEdit;
    btnSend: TButton;
    memResponse: TMemo;
    clAsyncClient1: TclAsyncClient;
    procedure btnConnectClick(Sender: TObject);
    procedure btnDisconnectClick(Sender: TObject);
    procedure btnSendClick(Sender: TObject);
    procedure clAsyncClient1Connect(Sender: TObject; ANextAction: TclNetworkStreamAction);
    procedure clAsyncClient1Read(Sender: TObject);
  end;

var
  Form2: TForm2;

implementation

{$R *.dfm}

procedure TForm2.btnConnectClick(Sender: TObject);
begin
  if (clAsyncClient1.Active) then raise Exception.Create('Client already connected');

  clAsyncClient1.Server := edtServer.Text;
  clAsyncClient1.Port := StrToInt(edtPort.Text);
  clAsyncClient1.TimeOut := 10000;

  //is necessary when using self-signed certificate by server
  clAsyncClient1.CertificateFlags := [cfIgnoreCommonNameInvalid, cfIgnoreUnknownAuthority];

  //specifies TLS 1.0 protocols (also available SSL 2.0 and SSL 3.0)
  clAsyncClient1.TlsFlags := [tfUseTLS];

  //forces the component to start SSL negotiation immediately once connected
  clAsyncClient1.UseTls := True;

  clAsyncClient1.Open();

  Caption := 'SSL Client - Connected';
end;

procedure TForm2.clAsyncClient1Connect(Sender: TObject; ANextAction: TclNetworkStreamAction);
begin
  if (ANextAction = saWrite) then
  begin
    clAsyncClient1.WriteData(nil);
  end;
end;

procedure TForm2.clAsyncClient1Read(Sender: TObject);
var
  stream: TMemoryStream;
  len: Int64;
  s: TclString;
begin
  stream := TMemoryStream.Create();
  try
    case (clAsyncClient1.ReadData(stream)) of
      saWrite: clAsyncClient1.WriteData(nil);
      saNone:
      begin
        if (stream.Size > SizeOf(Int64)) then
        begin
          //read size of incoming data
          stream.Position := 0;
          stream.Read(len, SizeOf(len));

          Assert((stream.Size - stream.Position) = len, 'To be simple, this situation is not handled in this sample');

          SetLength(s, len);

          //copy data
          stream.Read(PclChar(s)^, len);

          memResponse.Lines.Add('Received from server: ' + s);
        end;
      end;
    end;
  finally
    stream.Free();
  end;
end;

procedure TForm2.btnDisconnectClick(Sender: TObject);
begin
  if (clAsyncClient1.Close() = saWrite) then
  begin
    clAsyncClient1.WriteData(nil);
  end;
  Caption := 'SSL Client';
end;

procedure TForm2.btnSendClick(Sender: TObject);
var
  stream: TStream;
  len: Int64;
  s: TclString;
begin
  stream := TMemoryStream.Create();
  try
    //write the size of data
    s := edtData.Text;
    len := Length(s);
    stream.Write(len, SizeOf(len));

    //write data
    stream.Write(PclChar(s)^, len);

    stream.Position := 0;
    clAsyncClient1.WriteData(stream);
  finally
    stream.Free();
  end;
end;

end.

File: vcl/AsyncClientServerSSL/CB2010/ClientSSL/Unit2.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Unit2.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "clAsyncClient"
#pragma resource "*.dfm"
TForm2 *Form2;
//---------------------------------------------------------------------------
__fastcall TForm2::TForm2(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TForm2::btnConnectClick(TObject *Sender)
{
  if (clAsyncClient1->Active) throw new Exception("Client already connected");

  clAsyncClient1->Server = edtServer->Text;
  clAsyncClient1->Port = StrToInt(edtPort->Text);
  clAsyncClient1->TimeOut = 10000;

  //is necessary when using self-signed certificate by server
  clAsyncClient1->CertificateFlags = TclCertificateVerifyFlags() << cfIgnoreCommonNameInvalid << cfIgnoreUnknownAuthority;

  //specifies TLS 1.0 protocols (also available SSL 2.0 and SSL 3.0)
  clAsyncClient1->TLSFlags = TclTlsFlags() << tfUseTLS;

  //forces the component to start SSL negotiation immediately once connected
  clAsyncClient1->UseTLS = true;

  clAsyncClient1->Open();

  Caption = "SSL Client - Connected";
}
//---------------------------------------------------------------------------

void __fastcall TForm2::btnDisconnectClick(TObject *Sender)
{
  if (clAsyncClient1->Close() == saWrite) {
    clAsyncClient1->WriteData(NULL);
  }
  Caption = "SSL Client";
}
//---------------------------------------------------------------------------

void __fastcall TForm2::btnSendClick(TObject *Sender)
{
  TStream *stream = new TMemoryStream();
  __try {
    //write the size of data
    AnsiString s = edtData->Text;
    __int64 len = s.Length();
    stream->Write(&len, sizeof(len));

    //write data
    stream->Write(reinterpret_cast<PAnsiChar *>(s.c_str()), len);

    stream->Position = 0;
    clAsyncClient1->WriteData(stream);
  }
  __finally {
    delete stream;
  }
}
//---------------------------------------------------------------------------

void __fastcall TForm2::clAsyncClient1Connect(TObject *Sender, TclNetworkStreamAction ANextAction)

{
  if (ANextAction == saWrite) {
    clAsyncClient1->WriteData(NULL);
  }
}
//---------------------------------------------------------------------------

void __fastcall TForm2::clAsyncClient1Read(TObject *Sender)
{
  TStream *stream = new TMemoryStream();
  __try {
        switch (clAsyncClient1->ReadData(stream)) {
            case saWrite: {
              clAsyncClient1->WriteData(NULL);
              break;
            }
            case saNone: {
                if (stream->Size > sizeof(__int64)) {
                    //read size of incoming data
                    stream->Position = 0;

                    __int64 len;
                    stream->Read(&len, sizeof(len));

                    assert((stream->Size - stream->Position) == len); //To be simple, this situation is not handled in this sample

                    AnsiString s;
                    s.SetLength(len);

                    //copy data
                    stream->Read(reinterpret_cast<PAnsiChar *>(s.c_str()), len);

                    memResponse->Lines->Add("Received from server: " + s);
                }
                break;
            }
        }
  }
  __finally {
    delete stream;
  }
}
//---------------------------------------------------------------------------
